File: bindings/dart/test/handle_identity_test.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:maplibre_native_ffi/maplibre_native_ffi.dart';
import 'package:maplibre_native_ffi/src/internal/c/maplibre_native_c.dart'
    show threadLastErrorMessage;
import 'package:maplibre_native_ffi/src/runtime/runtime.dart'
    show mapHandleIdForTesting;
import 'package:maplibre_native_ffi/src/internal/c/maplibre_native_c.g.dart'
    as raw;
import 'package:maplibre_native_ffi/src/internal/status/status.dart';
import 'package:test/test.dart';

void _discardCompletion(
  Pointer<Void> _,
  Pointer<raw.mln_completion_result> _,
) {}

void _discardCompletionState(Pointer<Void> _) {}

/// Calls the C snapshot accessor with a raw map ID, so a test can replay a
/// released ID or use one from another isolate. The safe API expresses
/// neither.
void _mapSnapshotById(int map) {
  final arena = Arena();
  try {
    final snapshot = arena<raw.mln_map_snapshot>();
    snapshot.ref.size = sizeOf<raw.mln_map_snapshot>();
    checkNativeStatus(
      raw.mln_map_snapshot_get(map, snapshot),
      threadLastErrorMessage,
    );
  } finally {
    arena.releaseAll();
  }
}

void _runtimeBarrierById(int handle) {
  final arena = Arena();
  try {
    final completion = raw.mln_completion.$allocate(
      arena,
      size: sizeOf<raw.mln_completion>(),
      callback: Pointer.fromFunction<raw.mln_completion_callbackFunction>(
        _discardCompletion,
      ),
      user_data: nullptr,
      release_user_data:
          Pointer.fromFunction<raw.mln_completion_releaseFunction>(
            _discardCompletionState,
          ),
    );
    checkNativeStatus(
      raw.mln_runtime_barrier(handle, completion),
      threadLastErrorMessage,
    );
  } finally {
    arena.releaseAll();
  }
}

void main() {
  test(
    'a released map id replayed after a new map is reported stale',
    () async {
      final runtime = RuntimeHandle.create();
      final first = await runtime.createMap();
      final released = mapHandleIdForTesting(first);
      await first.close();

      // The released slot is the one the next map takes, so the replayed id
      // names a retired generation of a slot that is live again.
      final second = await runtime.createMap();
      addTearDown(() async {
        await second.close();
        await runtime.close();
      });

      expect(
        () => _mapSnapshotById(released),
        throwsA(
          isA<InvalidArgumentException>().having(
            (error) => error.diagnostic,
            'diagnostic',
            contains('stale'),
          ),
        ),
      );

      // The live map is unaffected by the replay.
      _mapSnapshotById(mapHandleIdForTesting(second));
    },
  );

  test(
    'a map id passed to a runtime operation is rejected on its kind',
    () async {
      final runtime = RuntimeHandle.create();
      final map = await runtime.createMap();
      addTearDown(() async {
        await map.close();
        await runtime.close();
      });

      // The generated bindings spell both as `int`, so this call needs the raw
      // id; the C API rejects it on its kind tag.
      expect(
        () => _runtimeBarrierById(mapHandleIdForTesting(map)),
        throwsA(
          isA<InvalidArgumentException>()
              .having((e) => e.diagnostic, 'diagnostic', contains('map'))
              .having((e) => e.diagnostic, 'diagnostic', contains('runtime')),
        ),
      );
    },
  );
}

File: bindings/go/handle_identity_test.go
package maplibre

import (
	"errors"
	"strings"
	"testing"
)

// BND-045: a released map id, replayed after a new map exists, is reported
// stale rather than naming the new map.
func TestReleasedMapIDReplayedAfterANewMapReportsItStale(t *testing.T) {
	runtime, err := NewRuntime()
	if err != nil {
		t.Fatalf("NewRuntime(): %v", err)
	}
	defer func() {
		if err := closeRuntimeForTest(runtime); err != nil {
			t.Errorf("Runtime Close(): %v", err)
		}
	}()

	first, err := awaitForTest(runtime.NewMap())
	if err != nil {
		t.Fatalf("NewMap(): %v", err)
	}
	released, err := first.ptr()
	if err != nil {
		t.Fatalf("ptr(): %v", err)
	}
	if err := closeMapForTest(first); err != nil {
		t.Fatalf("first.Close(): %v", err)
	}

	// The released slot is the one the next map takes, so the replayed id
	// names a retired generation of a slot that is live again.
	second, err := awaitForTest(runtime.NewMap())
	if err != nil {
		t.Fatalf("second NewMap(): %v", err)
	}
	defer func() { _ = closeMapForTest(second) }()

	err = mapSnapshotByIDForTest(released)
	if !errors.Is(err, ErrInvalidArgument) {
		t.Fatalf("replaying a released id: err = %v, want invalid argument", err)
	}
	if !strings.Contains(err.Error(), "stale") {
		t.Fatalf("diagnostic = %v, want it to name the id stale", err)
	}

	if _, err := second.Snapshot(); err != nil {
		t.Fatalf("second.Snapshot(): %v", err)
	}
}

// BND-047: a handle of one kind passed to another kind's entry point is
// rejected, and the diagnostic names both kinds.
func TestHandleOfAnotherKindIsRejectedByKind(t *testing.T) {
	_, m := newRuntimeAndMap(t, nil)

	raw, err := m.ptr()
	if err != nil {
		t.Fatalf("ptr(): %v", err)
	}
	err = runtimeEventMaskByIDForTest(nativeRuntime(raw))
	if !errors.Is(err, ErrInvalidArgument) {
		t.Fatalf("passing a map id to a runtime entry point: err = %v, want invalid argument", err)
	}
	diagnostic := err.Error()
	if !strings.Contains(diagnostic, "mln_map") || !strings.Contains(diagnostic, "mln_runtime") {
		t.Fatalf("diagnostic = %q, want it to name both handle kinds", diagnostic)
	}
}
